src/preprocess/visitor.ts

```diff
@@ -0,0 +1,702 @@
+      visitor.visitNode(resource.Overrides,
+        appendPath(path, `${name}.Overrides`),
+        // This is pre template expansion so the names
+        // used by $include, etc. must be in scope of the current
+        // environment, not in the template's env.
+        env));
+
+    // flag any names in the Template that should not be
+    // prefixed by visitRef, etc.
+    const $globalRefs: {[key: string]: boolean} = {};
+    _.forOwn(_.merge({}, resourceDoc.Parameters,
+      resourceDoc.Resources,
+      resourceDoc.Mappings,
+      resourceDoc.Conditions),
+      (param, key) => {
+        if (param.$global) {
+          $globalRefs[key] = true;
+        }
+      });
+
+    const $paramDefaultsEnv = mkSubEnv(
+      env, _.merge({Prefix: prefix}, template.$envValues), stackFrame);
+
+    const $paramDefaults = _.fromPairs(
+      _.filter(
+        _.map(
+          template.$params,
+          (v) => [v.Name,
+          visitor.visitNode(v.Default, appendPath(path, `${name}.$params.${v.Name}`), $paramDefaultsEnv)]),
+        ([k, v]) => !_.isUndefined(v)));
+
+    const providedParams = visitor.visitNode(resource.Properties, appendPath(path, `${name}.Properties`), env);
+    // TODO factor this out:
+    // TODO validate providedParams against template.$params[].type json-schema
+    // ! 1 find any missing params with no defaults
+    // 2 check against AllowedValues and AllowedPattern
+    // 3 check min/max Value / Length
+    const mergedParams = _.assign({}, $paramDefaults, providedParams);
+    _.forEach(template.$params, (param) => validateTemplateParameter(param, mergedParams, name, env));
+
+    const subEnv = mkSubEnv(
+      env,
+      _.merge(
+        {Prefix: prefix, $globalRefs},
+        mergedParams,
+        template.$envValues),
+      stackFrame);
+
+    // TODO consider just visitNode on the entire resourceDoc here
+    //      ... that requires finding a way to avoid double processing of .Resources
+    const outputResources = visitor.visitNode(resourceDoc.Resources, appendPath(path, `${name}.Resources`), subEnv)
+    // this will call visitNode on each section as it goes. See above ^
+    mapCustomResourceToGlobalSections(resourceDoc, path, subEnv);
+
+    // TODO allow individual output resources to have distinct $namePrefixes
+    //    prefix could be a map of resname: prefix
+    // Could also add a special char prefix individual resource names and global sections to
+    // override this name remapping.
+    // This ties in with supporting singleton custom resources that should be shared amongst templates
+    return _.map(_.toPairs(outputResources), ([resname, val]: [string, any]) => {
+      const isGlobal = _.has(val, '$global');
+      delete val.$global;
+      if (isGlobal) {
+        return [resname, val];
+      } else {
+        return [`${subEnv.$envValues.Prefix}${resname}`, val];
+      }
+    });
+  }
+
+}
+
+export class VariablesVisitor extends Visitor {
+  public variables: string[] = [];
```

Comment:
Is this just for testing purposes? How about `VariableTrackingVisitor`?

Reply:
`variables` is used in https://github.com/unbounce/iidy/blob/274f9508dd95fbb4f30c28dd321812b3369b7458/src/preprocess/filter.ts#L6-L19

I've renamed the class 👍